src/Controls/src/Core/TypeConversionHelper.cs

```diff
@@ -0,0 +1,172 @@
+#nullable enable
+
+using System;
+using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Maui.Controls.Xaml;
+using Microsoft.Maui.Hosting;
+
+namespace Microsoft.Maui.Controls
+{
+	internal static class TypeConversionHelper
+	{
+		internal static bool TryConvert(object value, Type targetType, out object? convertedValue)
+		{
+			Type valueType = value.GetType();
+
+			if (value is IWrappedValue { Value: var wrappedValue, ValueType: var wrappedType })
+			{
+				if (wrappedValue is null)
+				{
+					convertedValue = null;
+					return !targetType.IsValueType || targetType.IsGenericType && targetType.GetGenericTypeDefinition() == typeof(Nullable<>);
+				}
+
+				// It might be enough just to unwrap the value
+				if (targetType.IsAssignableFrom(wrappedType))
+				{
+					convertedValue = wrappedValue;
+					return true;
+				}
+
+				value = wrappedValue;
+				valueType = wrappedType;
+			}
+
+			if (TryGetTypeConverter(valueType, out var converter) && converter is not null && converter.CanConvertTo(targetType))
+			{
+				convertedValue = converter.ConvertTo(value, targetType) ?? throw new InvalidOperationException($"The {converter.GetType()} returned null when converting {valueType} to {targetType}");
+				return true;
+			}
+
+			if (TryGetTypeConverter(targetType, out converter) && converter is not null && converter.CanConvertFrom(valueType))
+			{
+				convertedValue = converter.ConvertFrom(value) ?? throw new InvalidOperationException($"The {converter.GetType()} returned null when converting from {valueType}");
+				return true;
+			}
+
+			if (RuntimeFeature.IsImplicitCastOperatorsUsageViaReflectionSupported)
+			{
+				if (TryConvertUsingImplicitCastOperator(value, targetType, out convertedValue))
+				{
+					return true;
+				}
+			}
+			else
+			{
+				WarnIfImplicitOperatorIsAvailable(value, targetType);
+			}
+
+			convertedValue = null;
+			return false;
+		}
+
+		private static bool TryGetTypeConverter(Type type, [NotNullWhen(true)] out TypeConverter? converter)
+			=> TypeConversionAppBuilderExtensions.TryGetTypeConverter(type, out converter)
+				|| type.TryGetTypeConverter(out converter);
+
+		[RequiresUnreferencedCode("The method uses reflection to find implicit conversion operators. " +
+			"It is not possible to guarantee that trimming does not remove some of the implicit operators. " +
+			"Consider preserving op_Implicit methods through DynamicDependency or DynamicallyAccessedMembers attributes.",
```

Comment:
Isn't the new suggestion to use the TypeConverter instead? I don't think we should actively ask people to use `DynamicDependency` as there's no way to validate that it's correct.